src/impl_/extract_argument.rs

```diff
@@ -1,10 +1,51 @@
 use crate::{
     exceptions::PyTypeError,
     ffi,
+    ffi_ptr_ext::FfiPtrExt,
     pyclass::boolean_struct::False,
-    types::{PyDict, PyString, PyTuple},
-    Bound, FromPyObject, PyAny, PyClass, PyErr, PyRef, PyRefMut, PyResult, PyTypeCheck, Python,
+    types::{any::PyAnyMethods, dict::PyDictMethods, tuple::PyTupleMethods, PyDict, PyTuple},
+    Borrowed, Bound, FromPyObject, PyAny, PyClass, PyErr, PyRef, PyRefMut, PyResult, PyTypeCheck,
+    Python,
 };
+use crate::{PyObject, ToPyObject};
+
+/// Function argument extraction borrows input arguments.
+///
+/// This might potentially be removed in favour of just using `Borrowed` directly in future,
+/// for now it's used just to control the traits and methods available while the `Bound` API
+/// is being rounded out.
+#[derive(Copy, Clone)]
+#[repr(transparent)]
+pub struct PyArg<'py>(Borrowed<'py, 'py, PyAny>);
+
+impl<'py> PyArg<'py> {
+    #[inline]
+    pub unsafe fn from_ptr_or_opt(py: Python<'py>, ptr: *mut ffi::PyObject) -> Option<Self> {
+        ptr.assume_borrowed_or_opt(py).map(Self)
+    }
+
+    #[inline]
+    pub unsafe fn from_ptr(py: Python<'py>, ptr: *mut ffi::PyObject) -> Self {
+        Self(ptr.assume_borrowed(py))
+    }
+
+    #[inline]
+    pub fn from_borrowed<T>(borrowed: Borrowed<'py, 'py, T>) -> Self {
+        Self(borrowed.into_any())
+    }
+
+    #[inline]
+    pub fn as_borrowed(self) -> Borrowed<'py, 'py, PyAny> {
+        self.0
+    }
+}
```

Comment:
To be more specific to answer the docstring's future possibility of just using `Borrowed` everywhere, I think that the only step needed is to expose `Borrowed::from_ptr` family of functions as public API. Maybe I should break that into a separate PR and then I could rework this PR one more time to avoid this newtype?

Reply:
I think we would need to expose them anyway, if we want to provide any replacement for the `Python::from_borrowed_ptr` family, right?

Reply:
I think the alternative would be to force users to increase reference counts and take ownership with `Bound::from_borrowed_ptr` methods instead, like we have on `Py`. There's a potentially good argument to do that for simplicity and consistency, and keep `Borrowed` deliberately minimalist.

Yet another option is to add both sets of methods and just encourage users to use `Bound` instead of `Borrowed` as much as possible.

Reply:
Ok, I've removed this type.